Support for managed pulling from private ECRs

Caomoji opened the pull request:
This PR adds the possibility to set additional references to ECRs as sources in the configuration file.

This change allows to take advantage of the logic already in place to rotate ECR tokens and authenticate to private registries, but for pulling images from private ECRs in a context of multi ECR replication. The same AWS credentials as for the target ECR are used. 

#### Main changes:
- Updated the configuration to add the field `privateRegistries` in `source`
- Added a new attribute to the `imagePullSecretProviders` containing the registries' clients
- Added a function to produce a `dockerconfig` in a JSON format from a registry client to merge with the authfile passed to Skopeo
- Updated `GetImagePullSecrets` to include dockerconfigs from private registries to the image pull secrets from pods
  
#### Notes:
- `imagePullSecrets` from hooked Pods still have priority over the authentication via these default private registries 
- Changes do not impact the Helm chart and are compatible with previous version
- Source private registries cannot authenticate with different credentials from the ones used by the target registry passed as environment variables

Caomoji commented:
Hello @estahn, 
Sorry for bothering, but we would like to ask if you are possibly interested in this PR, and if we could expect to have it eventually merged or if it's not a move you would like for this project.
I'm open to any change suggestions.
Thank you in advance 🙏

estahn commented:
> Hello @estahn, 
> 
> Sorry for bothering, but we would like to ask if you are possibly interested in this PR, and if we could expect to have it eventually merged or if it's not a move you would like for this project.
> 
> I'm open to any change suggestions.
> 
> Thank you in advance 🙏


PRs are much appreciated. Let me take a look at it today and come back to you shortly.

estahn commented:
@Caomoji I prepared this initially to allow for registries requiring authentication (https://github.com/estahn/k8s-image-swapper/blame/bcc56b57ac91c1e8312f4e558283c68684a38678/.k8s-image-swapper.yml#L34-L37).

I prefer not to distinguish between private & public registries, but rather have `registries` requiring additional configuration. I had requests to support Azure Image Registries and I plan at looking at this soon. It would be good to have this already incorporated:

My suggestion is to adjust this as follows:
```yaml
source:
  registries:
    - type: aws
      aws:
        accountId: 123456789
        region: ap-southeast-2
    - type: aws
      aws:
        accountId: 234567890
        region: us-east-1
    - type: dockerio
      dockerio:
        username: 
        password: 
        secretRef: 
```

Thoughts?

Caomoji commented:
> @Caomoji I prepared this initially to allow for registries requiring authentication ([`bcc56b5`/.k8s-image-swapper.yml#L34-L37 (blame)](https://github.com/estahn/k8s-image-swapper/blame/bcc56b57ac91c1e8312f4e558283c68684a38678/.k8s-image-swapper.yml#L34-L37)).
> 
> I prefer not to distinguish between private & public registries, but rather have `registries` requiring additional configuration. I had requests to support Azure Image Registries and I plan at looking at this soon. It would be good to have this already incorporated:
> 
> My suggestion is to adjust this as follows:
> 
> ```yaml
> source:
>   registries:
>     - type: aws
>       aws:
>         accountId: 123456789
>         region: ap-southeast-2
>     - type: aws
>       aws:
>         accountId: 234567890
>         region: us-east-1
>     - type: dockerio
>       dockerio:
>         username: 
>         password: 
>         secretRef: 
> ```
> 
> Thoughts?

I fully agree with your point of view, I updated the config to look like your example, as well as some methods to make registry client creation a bit cleaner. I also updated the target registry for consistency.
Adding other types of registry clients and implementing them should fit nicely now I think, please let me know you thoughts.

Caomoji commented:
I rebased on `main` and adapted some things:
- Integrated changes with the new `ECROptions` field in AWS config
- Removed function `SetRepositoryTags` specific to ECR to rather integrate tags & options as part of the Client creation
- Made function `newECRClient` internal to restrict its use and rather emphasis on the more generic `NewClient`

@estahn please let me know if you have any suggestions.

estahn commented on cmd/root.go:
```diff
@@ -63,14 +63,24 @@ A mutating webhook for Kubernetes, pointing the images to a new location.`,
+			sourceRegistryClient, err := registry.NewClient(reg)
+			if err != nil {
+				log.Err(err).Msg(fmt.Sprintf("error connecting to source registry at %s", reg.GetServerAddress()))
+				os.Exit(1)
```
I wonder if this is being too aggressive. If there are multiple sources and one fails it stops everything? Is that your expected failure mode?

Caomoji replied:
In our use case, properly connecting to source registries is mandatory; while writing this we had in mind that if we pass registries in the configuration, they should work and that if they don't, then it's preferable to abort everything rather than to start, thinking we passed some configuration which actually didn't apply because of the initial connection failure.
However I agree that it should not always be mandatory, maybe we could add a mechanism to retry the connection in the background in case of error, so image-swapper can start as usual in the meantime and eventually match its configuration as the connection to the registry.ies is restored and added back to the list of authenticated registries. 
What do you think?

estahn replied:
I can see the advantage of failing early. I don't think we necessarily need a retry mechanism here given the pod will be restarted by Kubernetes anyways. If this becomes an issue we can address it separately.

Caomoji commented:
Hello @estahn,
Do you think we could go on with this PR?
I'm available for any change suggestions, please let me know :pray:

Caomoji commented:
@estahn Thank you for your response, no worries.

The two PRs were quite conflicting indeed, both trying to bring genericity to registry handling but in a different way. I resolved the conflicts by trying to take the best of the two, while taking into account the core changes of this PR (managing sources registries as well).
I especially tried to make ECR and GAR registry clients look the same in their respective implementation, I modified some attributes from GAR client to look like ECR and rationalized common stuff.

estahn commented:
> @estahn Thank you for your response, no worries.
> 
> The two PRs were quite conflicting indeed, both trying to bring genericity to registry handling but in a different way. I resolved the conflicts by trying to take the best of the two, while taking into account the core changes of this PR (managing sources registries as well). I especially tried to make ECR and GAR registry clients look the same in their respective implementation, I modified some attributes from GAR client to look like ECR and rationalized common stuff.

Awesome. Will check it out tonight. Did you have a chance to run it locally to test it?

estahn commented on docs/configuration.md:
```diff
@@ -155,12 +182,13 @@ It's a slice of `Key` and `Value`.
-          tags:
-            - key: cluster
-              value: myCluster
+      registry:
```
@Caomoji Does `target` have other attributes besides `registry`? If not, I would like to avoid having another level underneath because:
* simplicity, avoiding nesting if not strictly necessary
* backwards compatibility

Let me know your thoughts.

Would it help to change the type of `Target` to `Registry`?

https://github.com/estahn/k8s-image-swapper/blob/78f64244d77aaab14f78ef82174094616f4c9c77/pkg/config/config.go#L45

Example:
```
Target Registry `yaml:"target"`
```

Caomoji replied:
As of now I agree it makes more sense to directly set the registry fields under `target`, I did the update :+1:

estahn commented on docs/configuration.md:
```diff
@@ -171,9 +199,10 @@ The GCP location, projectId, and repositoryId are used to constrct the GCP Artif
-        location: us-central1
-        projectId: gcp-project-123
-        repositoryId: main
+      registry:
```
Same as above.
